File: 2026-01/src/day_13/boj_11444/Main.cpp
#include <bits/stdc++.h>
using namespace std;

using Matrix = vector<vector<long long>>;

constexpr int MOD = 1000000007;

Matrix identity() {
    return {{1, 0}, {0, 1}};
}

Matrix base_matrix() {
    return {{1, 1}, {1, 0}};
}

Matrix multiply(const Matrix& m1, const Matrix& m2) {
    return {{(m1[0][0] * m2[0][0] + m1[0][1] * m2[1][0]) % MOD, (m1[0][0] * m2[0][1] + m1[0][1] * m2[1][1]) % MOD},
            {(m1[1][0] * m2[0][0] + m1[1][1] * m2[1][0]) % MOD, (m1[1][0] * m2[0][1] + m1[1][1] * m2[1][1]) % MOD}};
}

Matrix binpow(Matrix mat, long long n) {
    Matrix res = identity();

    while (n > 0) {
        if (n & 1) res = multiply(res, mat);
        mat = multiply(mat, mat);
        n >>= 1;
    }

    return res;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    long long n;
    cin >> n;

    if (n == 0) {
        cout << 0;
        return 0;
    }

    Matrix res = binpow(base_matrix(), n - 1);
    cout << res[0][0];
}

File: 2026-01/src/day_13/boj_11444/Main.java
package day_13.boj_11444;

import java.io.*;

public class Main {

    static final int MOD = 1_000_000_007;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        long N = Long.parseLong(br.readLine());
        if (N == 0) {
            System.out.println(0);
            return;
        }

        long[][] res = binPow(baseMatrix(), N - 1);
        System.out.println(res[0][0]);
    }

    static long[][] binPow(long[][] mat, long n) {
        long[][] res = identity();

        while (n > 0) {
            if ((n & 1) > 0) res = multiply(res, mat);
            mat = multiply(mat, mat);
            n >>= 1;
        }

        return res;
    }

    static long[][] identity() {
        return new long[][]{
                {1, 0},
                {0, 1},
        };
    }

    static long[][] baseMatrix() {
        return new long[][]{
                {1, 1},
                {1, 0},
        };
    }

    static long[][] multiply(long[][] m1, long[][] m2) {
        return new long[][]{
                {(m1[0][0] * m2[0][0] + m1[0][1] * m2[1][0]) % MOD, (m1[0][0] * m2[0][1] + m1[0][1] * m2[1][1]) % MOD},
                {(m1[1][0] * m2[0][0] + m1[1][1] * m2[1][0]) % MOD, (m1[1][0] * m2[0][1] + m1[1][1] * m2[1][1]) % MOD}
        };
    }
}
